File: core/src/main/java/dev/vibris/core/FixedShaderLink.java
package dev.vibris.core;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.attribute.BasicFileAttributes;
import java.nio.file.attribute.FileTime;
import java.util.Objects;

import static java.nio.file.LinkOption.NOFOLLOW_LINKS;

final class FixedShaderLink implements ShaderLink {
    private static final String ACTIVE_NAME = "shaders";
    private static final String NEXT_PREFIX = "shaders.next.";

    private final Path pendingRoot;
    private final Path shaderpackRoot;
    private final Path activeLink;
    private final AtomicMover atomicMover;

    FixedShaderLink(Path pendingRoot, Path shaderpackRoot) {
        this(pendingRoot, shaderpackRoot, (source, target) -> Files.move(
            source, target, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING));
    }

    FixedShaderLink(Path pendingRoot, Path shaderpackRoot, AtomicMover atomicMover) {
        this.pendingRoot = pendingRoot.toAbsolutePath().normalize();
        this.shaderpackRoot = shaderpackRoot.toAbsolutePath().normalize();
        activeLink = this.shaderpackRoot.resolve(ACTIVE_NAME);
        this.atomicMover = Objects.requireNonNull(atomicMover, "atomicMover");
    }

    synchronized void prepare() throws Failure {
        requireOrdinaryRoots();
        detach();
        try (var children = Files.list(shaderpackRoot)) {
            for (Path child : children.filter(path -> path.getFileName().toString().startsWith(NEXT_PREFIX)).toList()) {
                deleteLink(child);
            }
        } catch (IOException exception) {
            throw new Failure("Failed to clean temporary shader links.", true, exception);
        }
    }

    @Override
    public synchronized void switchTo(SourceRegistry.Lease source, OwnershipCheck ownership) throws Failure {
        requireOrdinaryRoots();
        Path target = source.directory().toAbsolutePath().normalize();
        if (!pendingRoot.equals(target.getParent())) {
            throw new Failure("Shader source is outside the pending root.", true);
        }
        requireLinkOrMissing(activeLink);
        Path temporary = shaderpackRoot.resolve(NEXT_PREFIX + source.uuid());
        if (Files.exists(temporary, NOFOLLOW_LINKS)) {
            throw new Failure("Temporary shader link already exists.", true);
        }

        LinkIdentity identity = null;
        try {
            Files.createSymbolicLink(temporary, target);
            identity = LinkIdentity.capture(temporary);
            ownership.verify();
            if (!identity.matches(temporary) || !target.equals(Files.readSymbolicLink(temporary))) {
                throw new Failure("Temporary shader link identity changed.", true);
            }
            replace(temporary, target);
        } catch (Failure failure) {
            deleteIfOwned(temporary, identity);
            throw failure;
        } catch (IOException | UnsupportedOperationException | SecurityException exception) {
            deleteIfOwned(temporary, identity);
            throw new Failure("Failed to create the directory shader link.", true, exception);
        }
    }

    @Override
    public synchronized void detach() throws Failure {
        requireLinkOrMissing(activeLink);
        if (!Files.exists(activeLink, NOFOLLOW_LINKS)) return;
        try {
            Files.delete(activeLink);
        } catch (IOException exception) {
            throw new Failure("Failed to detach the active shader link.", true, exception);
        }
    }

    @Override
    public boolean retainsActiveSource() {
        return true;
    }

    synchronized Path currentTarget() throws Failure {
        requireLinkOrMissing(activeLink);
        if (!Files.exists(activeLink, NOFOLLOW_LINKS)) return null;
        try {
            return Files.readSymbolicLink(activeLink);
        } catch (IOException exception) {
            throw new Failure("Failed to read the active shader link.", false, exception);
        }
    }

    private void replace(Path temporary, Path target) throws Failure {
        Path previous = currentTarget();
        try {
            atomicMover.move(temporary, activeLink);
        } catch (IOException atomicFailure) {
            replaceGuarded(temporary, previous, atomicFailure);
        }
        try {
            if (!Files.isSymbolicLink(activeLink) || !target.equals(Files.readSymbolicLink(activeLink))) {
                throw new IOException("active shader link target mismatch");
            }
        } catch (IOException verificationFailure) {
            boolean restored = restore(previous);
            throw new Failure("Active shader link verification failed.", restored, verificationFailure);
        }
    }

    private void replaceGuarded(Path temporary, Path previous, IOException atomicFailure) throws Failure {
        requireLinkOrMissing(activeLink);
        try {
            if (Files.exists(activeLink, NOFOLLOW_LINKS)) Files.delete(activeLink);
            Files.move(temporary, activeLink);
        } catch (IOException fallbackFailure) {
            fallbackFailure.addSuppressed(atomicFailure);
            boolean restored = restore(previous);
            throw new Failure("Failed to replace the active shader link.", restored, fallbackFailure);
        }
    }

    private boolean restore(Path previous) {
        try {
            if (Files.exists(activeLink, NOFOLLOW_LINKS)) deleteLink(activeLink);
            if (previous != null) Files.createSymbolicLink(activeLink, previous);
            return previous == null
                ? !Files.exists(activeLink, NOFOLLOW_LINKS)
                : Files.isSymbolicLink(activeLink) && previous.equals(Files.readSymbolicLink(activeLink));
        } catch (IOException | Failure exception) {
            return false;
        }
    }

    private void requireOrdinaryRoots() throws Failure {
        try {
            requireOrdinaryAncestors(pendingRoot);
            requireOrdinaryAncestors(shaderpackRoot);
        } catch (IOException exception) {
            throw new Failure("Shader link roots are missing or unsafe.", true, exception);
        }
    }

    private static void requireOrdinaryAncestors(Path path) throws IOException {
        Path current = path;
        while (current != null) {
            OwnedPathIdentity.captureDirectory(current);
            current = current.getParent();
        }
    }

    private static void requireLinkOrMissing(Path path) throws Failure {
        if (Files.exists(path, NOFOLLOW_LINKS) && !Files.isSymbolicLink(path)) {
            throw new Failure("Shader link path is not a symbolic link.", true);
        }
    }

    private static void deleteLink(Path path) throws IOException, Failure {
        requireLinkOrMissing(path);
        if (Files.exists(path, NOFOLLOW_LINKS)) Files.delete(path);
    }

    private static void deleteIfOwned(Path path, LinkIdentity identity) {
        if (identity == null || !identity.matches(path)) return;
        try {
            Files.delete(path);
        } catch (IOException ignored) {
        }
    }

    @FunctionalInterface
    interface AtomicMover {
        void move(Path source, Path target) throws IOException;
    }

    private record LinkIdentity(Object fileKey, FileTime creationTime, Path target) {
        static LinkIdentity capture(Path path) throws IOException {
            BasicFileAttributes attributes = Files.readAttributes(path, BasicFileAttributes.class, NOFOLLOW_LINKS);
            if (!attributes.isSymbolicLink() || !Files.isSymbolicLink(path)) {
                throw new IOException("path is not a symbolic link");
            }
            return new LinkIdentity(attributes.fileKey(), attributes.creationTime(), Files.readSymbolicLink(path));
        }

        boolean matches(Path path) {
            try {
                LinkIdentity current = capture(path);
                boolean sameFile = fileKey != null || current.fileKey != null
                    ? fileKey != null && fileKey.equals(current.fileKey)
                    : creationTime.equals(current.creationTime);
                return sameFile && target.equals(current.target);
            } catch (IOException exception) {
                return false;
            }
        }
    }
}

File: core/src/main/kotlin/dev/vibris/core/FixedShaderLink.kt
package dev.vibris.core

import java.io.IOException
import java.nio.file.Files
import java.nio.file.LinkOption.NOFOLLOW_LINKS
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.attribute.BasicFileAttributes
import java.nio.file.attribute.FileTime
import java.util.Objects

internal class FixedShaderLink(
    pendingRoot: Path,
    shaderpackRoot: Path,
    atomicMover: AtomicMover,
) : ShaderLink {
    private val pendingRoot = pendingRoot.toAbsolutePath().normalize()
    private val shaderpackRoot = shaderpackRoot.toAbsolutePath().normalize()
    private val activeLink = this.shaderpackRoot.resolve(ACTIVE_NAME)
    private val atomicMover = Objects.requireNonNull(atomicMover, "atomicMover")
    private var activeIdentity: LinkIdentity? = null

    constructor(pendingRoot: Path, shaderpackRoot: Path) : this(
        pendingRoot,
        shaderpackRoot,
        AtomicMover { source, target ->
            Files.move(
                source,
                target,
                StandardCopyOption.ATOMIC_MOVE,
                StandardCopyOption.REPLACE_EXISTING,
            )
            Unit
        },
    )

    @Synchronized
    @Throws(ShaderLink.Failure::class)
    fun prepare() {
        requireOrdinaryRoots()
        try {
            deleteLink(activeLink)
            activeIdentity = null
        } catch (exception: IOException) {
            throw ShaderLink.Failure("Failed to detach the stale shader link.", true, exception)
        }
        try {
            Files.list(shaderpackRoot).use { children ->
                for (child in children.filter { path -> path.fileName.toString().startsWith(NEXT_PREFIX) }.toList()) {
                    deleteLink(child)
                }
            }
        } catch (exception: IOException) {
            throw ShaderLink.Failure("Failed to clean temporary shader links.", true, exception)
        }
    }

    @Synchronized
    @Throws(ShaderLink.Failure::class)
    override fun switchTo(source: SourceRegistry.Lease, ownership: ShaderLink.OwnershipCheck) {
        requireOrdinaryRoots()
        verifyActiveLink()
        val target = source.directory.toAbsolutePath().normalize()
        if (pendingRoot != target.parent) {
            throw ShaderLink.Failure("Shader source is outside the pending root.", true)
        }
        requireLinkOrMissing(activeLink)
        val temporary = shaderpackRoot.resolve(NEXT_PREFIX + source.uuid)
        if (Files.exists(temporary, NOFOLLOW_LINKS)) {
            throw ShaderLink.Failure("Temporary shader link already exists.", true)
        }

        var identity: LinkIdentity? = null
        try {
            Files.createSymbolicLink(temporary, target)
            identity = LinkIdentity.capture(temporary)
            ownership.verify()
            if (!identity.matches(temporary) || target != Files.readSymbolicLink(temporary)) {
                throw ShaderLink.Failure("Temporary shader link identity changed.", true)
            }
            replace(temporary, target)
        } catch (failure: ShaderLink.Failure) {
            deleteIfOwned(temporary, identity)
            throw failure
        } catch (exception: IOException) {
            deleteIfOwned(temporary, identity)
            throw ShaderLink.Failure("Failed to create the directory shader link.", true, exception)
        } catch (exception: UnsupportedOperationException) {
            deleteIfOwned(temporary, identity)
            throw ShaderLink.Failure("Failed to create the directory shader link.", true, exception)
        } catch (exception: SecurityException) {
            deleteIfOwned(temporary, identity)
            throw ShaderLink.Failure("Failed to create the directory shader link.", true, exception)
        }
    }

    @Synchronized
    @Throws(ShaderLink.Failure::class)
    override fun detach() {
        try {
            verifyActiveLink()
        } catch (failure: ShaderLink.Failure) {
            if (Files.isSymbolicLink(activeLink)) {
                try {
                    Files.delete(activeLink)
                    activeIdentity = null
                } catch (cleanupFailure: IOException) {
                    failure.addSuppressed(cleanupFailure)
                }
            }
            throw failure
        }
        if (activeIdentity == null) {
            return
        }
        try {
            Files.delete(activeLink)
            activeIdentity = null
        } catch (exception: IOException) {
            throw ShaderLink.Failure("Failed to detach the active shader link.", true, exception)
        }
    }

    @Synchronized
    @Throws(ShaderLink.Failure::class)
    override fun retainsActiveSource(): Boolean {
        verifyActiveLink()
        return activeIdentity != null
    }

    @Synchronized
    @Throws(ShaderLink.Failure::class)
    fun currentTarget(): Path? {
        verifyActiveLink()
        return activeIdentity?.target
    }

    @Throws(ShaderLink.Failure::class)
    private fun replace(temporary: Path, target: Path) {
        val previous = currentTarget()
        try {
            atomicMover.move(temporary, activeLink)
        } catch (atomicFailure: IOException) {
            replaceGuarded(temporary, previous, atomicFailure)
        }
        try {
            val replacement = LinkIdentity.capture(activeLink)
            if (target != replacement.target) {
                throw IOException("active shader link target mismatch")
            }
            activeIdentity = replacement
        } catch (verificationFailure: IOException) {
            val restored = restore(previous)
            throw ShaderLink.Failure("Active shader link verification failed.", restored, verificationFailure)
        }
    }

    @Throws(ShaderLink.Failure::class)
    private fun replaceGuarded(temporary: Path, previous: Path?, atomicFailure: IOException) {
        verifyActiveLink()
        try {
            if (Files.exists(activeLink, NOFOLLOW_LINKS)) {
                Files.delete(activeLink)
            }
            Files.move(temporary, activeLink)
        } catch (fallbackFailure: IOException) {
            fallbackFailure.addSuppressed(atomicFailure)
            val restored = restore(previous)
            throw ShaderLink.Failure("Failed to replace the active shader link.", restored, fallbackFailure)
        }
    }

    private fun restore(previous: Path?): Boolean {
        return try {
            if (Files.exists(activeLink, NOFOLLOW_LINKS)) {
                deleteLink(activeLink)
            }
            if (previous != null) {
                Files.createSymbolicLink(activeLink, previous)
            }
            activeIdentity = if (previous == null) null else LinkIdentity.capture(activeLink)
            previous == null || previous == checkNotNull(activeIdentity).target
        } catch (_: IOException) {
            activeIdentity = null
            false
        } catch (_: ShaderLink.Failure) {
            activeIdentity = null
            false
        }
    }

    @Throws(ShaderLink.Failure::class)
    private fun verifyActiveLink() {
        val identity = activeIdentity
        if (identity == null) {
            if (Files.exists(activeLink, NOFOLLOW_LINKS)) {
                throw ShaderLink.Failure("Active shader link was created outside Vibris.", false)
            }
            return
        }
        if (!identity.matches(activeLink)) {
            throw ShaderLink.Failure("Active shader link identity or target changed.", false)
        }
    }

    @Throws(ShaderLink.Failure::class)
    private fun requireOrdinaryRoots() {
        try {
            requireOrdinaryAncestors(pendingRoot)
            requireOrdinaryAncestors(shaderpackRoot)
        } catch (exception: IOException) {
            throw ShaderLink.Failure("Shader link roots are missing or unsafe.", true, exception)
        }
    }

    fun interface AtomicMover {
        @Throws(IOException::class)
        fun move(source: Path, target: Path)
    }

    private data class LinkIdentity(
        val fileKey: Any?,
        val creationTime: FileTime,
        val target: Path,
    ) {
        fun matches(path: Path): Boolean {
            return try {
                val current = capture(path)
                val sameFile = if (fileKey != null || current.fileKey != null) {
                    fileKey != null && fileKey == current.fileKey
                } else {
                    creationTime == current.creationTime
                }
                sameFile && target == current.target
            } catch (_: IOException) {
                false
            }
        }

        companion object {
            @Throws(IOException::class)
            fun capture(path: Path): LinkIdentity {
                val attributes = Files.readAttributes(path, BasicFileAttributes::class.java, NOFOLLOW_LINKS)
                if (!attributes.isSymbolicLink || !Files.isSymbolicLink(path)) {
                    throw IOException("path is not a symbolic link")
                }
                return LinkIdentity(attributes.fileKey(), attributes.creationTime(), Files.readSymbolicLink(path))
            }
        }
    }

    companion object {
        private const val ACTIVE_NAME = "shaders"
        private const val NEXT_PREFIX = "shaders.next."

        @Throws(IOException::class)
        private fun requireOrdinaryAncestors(path: Path) {
            var current: Path? = path
            while (current != null) {
                OwnedPathIdentity.captureDirectory(current)
                current = current.parent
            }
        }

        @Throws(ShaderLink.Failure::class)
        private fun requireLinkOrMissing(path: Path) {
            if (Files.exists(path, NOFOLLOW_LINKS) && !Files.isSymbolicLink(path)) {
                throw ShaderLink.Failure("Shader link path is not a symbolic link.", true)
            }
        }

        @Throws(IOException::class, ShaderLink.Failure::class)
        private fun deleteLink(path: Path) {
            requireLinkOrMissing(path)
            if (Files.exists(path, NOFOLLOW_LINKS)) {
                Files.delete(path)
            }
        }

        private fun deleteIfOwned(path: Path, identity: LinkIdentity?) {
            if (identity == null || !identity.matches(path)) {
                return
            }
            try {
                Files.delete(path)
            } catch (_: IOException) {
            }
        }
    }
}